src/main/java/de/tum/cit/aet/artemis/programming/service/localvc/LocalVCPrePushHook.java

```diff
@@ -73,10 +73,21 @@ public void onPreReceive(ReceivePack receivePack, Collection<ReceiveCommand> com
             return;
         }
 
-        // Reject pushes to anything other than the default branch.
-        if (!command.getRefName().equals("refs/heads/" + defaultBranchName)) {
-            command.setResult(ReceiveCommand.Result.REJECTED_OTHER_REASON, "You cannot push to a branch other than the default branch.");
-            return;
+        // If branching is allowed, reject pushes to anything other than branch names that match the branch regex
+        // If branching is not allowed, reject pushes to anything other than the default branch
+        boolean isBranchingAllowed = localVCServletService.isBranchingAllowedForRepository(repository);
+        String branchPath = command.getRefName();
+        if (!branchPath.equals("refs/heads/" + defaultBranchName)) {
+            String branchName = branchPath.substring("refs/heads/".length());
+            if (!isBranchingAllowed) {
+                command.setResult(ReceiveCommand.Result.REJECTED_OTHER_REASON, "You cannot push to a branch other than the default branch.");
+                return;
+            }
+            else if (!localVCServletService.isBranchNameAllowedForRepository(repository, branchName)) {
```

Comment:
I think you are fetching the same information (exercise+build config) from the database twice here, once in `isBranchingAllowed` and then again in `isBranchNameAllowed`.